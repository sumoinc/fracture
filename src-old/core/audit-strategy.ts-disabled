import { ResourceAttributeOptions } from "./resource-attribute";

export type AuditStrategy = {
  create: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
  update: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
  // turning this on activates soft deletes
  delete: {
    dateAttribute?: ResourceAttributeOptions;
    userAttribute?: ResourceAttributeOptions;
  };
};
