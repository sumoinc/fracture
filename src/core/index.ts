export * from "./access-pattern";
export * from "./account";
export * from "./audit-strategy";
export * from "./fracture";
export * from "./index";
export * from "./naming-strategy";
export * from "./operation";
export * from "./organization";
export * from "./organizational-unit";
export * from "./region";
export * from "./resource-attribute";
export * from "./resource";
export * from "./service";
export * from "./structure";
