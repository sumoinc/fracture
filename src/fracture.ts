import { Project } from 'projen';
import { AppSync } from './appsync/appsync';
import { FractureComponent } from './component';
import { Entity, EntityOptions } from './model';

export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   */
  gendir?: string;
  /**
   * Enable AppSync
   */
  appsync?: boolean;
}

export class Fracture extends FractureComponent {
  public readonly gendir: string;
  public readonly appsync: boolean;

  constructor(
    project: Project,
    namespace?: string,
    options: FractureOptions = {},
  ) {
    super(project, namespace ?? 'fracture' );

    this.gendir = options.gendir ?? project.outdir;
    this.appsync = options.appsync ?? true;

    if (this.appsync) {
      new AppSync(project, this.namespace, { gendir: this.gendir });
    }
  }

  /**
   * Get all entities in this project's namespace.
   */
  public get entities(): Entity[] {
    const isEntity = (c: FractureComponent): c is Entity => c instanceof Entity && c.namespace === this.namespace;
    return (this.project.components as FractureComponent[]).filter(isEntity);
  }

  public addEntity(options: EntityOptions) {
    return new Entity(this.project, this.namespace, options);
  }


}
