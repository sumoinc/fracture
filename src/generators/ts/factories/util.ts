import { ValueOf } from "type-fest";
import { ResourceAttributeType } from "../../../services";

export const JsType = {
  STRING: "string",
  NUMBER: "number",
  BOOLEAN: "boolean",
  ARRAY: "array",
  MAP: "map",
  OTHER: "other",
} as const;

export const jsType = (
  resourceAttributeType: ValueOf<typeof ResourceAttributeType> | string
) => {
  switch (resourceAttributeType) {
    case ResourceAttributeType.GUID:
    case ResourceAttributeType.STRING:
    case ResourceAttributeType.EMAIL:
    case ResourceAttributeType.PHONE:
    case ResourceAttributeType.URL:
    case ResourceAttributeType.DATE:
    case ResourceAttributeType.TIME:
    case ResourceAttributeType.DATE_TIME:
    case ResourceAttributeType.JSON:
    case ResourceAttributeType.IPADDRESS:
      return JsType.STRING;
    case ResourceAttributeType.INT:
    case ResourceAttributeType.FLOAT:
    case ResourceAttributeType.TIMESTAMP:
    case ResourceAttributeType.COUNT:
    case ResourceAttributeType.AVERAGE:
    case ResourceAttributeType.SUM:
      return JsType.NUMBER;
    case ResourceAttributeType.BOOLEAN:
      return JsType.BOOLEAN;
    case ResourceAttributeType.ARRAY:
      return JsType.ARRAY;
    case ResourceAttributeType.MAP:
      return JsType.MAP;
    default:
      return JsType.OTHER;
  }
};
