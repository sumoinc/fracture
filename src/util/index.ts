export * from "./determineBuildId";
export * from "./test-util";
