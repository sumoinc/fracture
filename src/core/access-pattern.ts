import { deepMerge } from "projen/lib/util";
import { ValueOf } from "type-fest";
import { Operation, OPERATION_SUB_TYPE } from "./operation";
import { Resource } from "./resource";
import {
  ResourceAttribute,
  ResourceAttributeGenerator,
  ResourceAttributeOptions,
} from "./resource-attribute";
import { FractureComponent } from "../core/component";
import { DynamoGsi } from "../dynamodb/dynamo-gsi";
import { DynamoTable } from "../dynamodb/dynamo-table";

export const ACCESS_PATTERN_TYPE = {
  /**
   * Main PK and PK pattern used to identify this record.
   */
  IDENTIFIER: "Identifier",
  /**
   * Lookup pattern for quick begins_with() searching and such.
   */
  LOOKUP: "lookup",
  /**
   * Any other AP
   */
  OTHER: "other",
} as const;

export interface AccessPatternOptions {
  dynamoGsi?: DynamoGsi;
  pkAttributeOptions?: ResourceAttributeOptions;
  skAttributeOptions?: ResourceAttributeOptions;
  type?: ValueOf<typeof ACCESS_PATTERN_TYPE>;
}

export class AccessPattern extends FractureComponent {
  // member components
  public pkAttribute: ResourceAttribute;
  public skAttribute: ResourceAttribute;
  public operations: Operation[] = [];
  // parent
  public readonly resource: Resource;
  // all other options
  public readonly options: Required<AccessPatternOptions>;

  constructor(resource: Resource, options: AccessPatternOptions) {
    super(resource.fracture);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     **************************************************************************/

    // make a new gsi is one wasn't passed into the Access Pattern
    const dynamoGsi = options.dynamoGsi
      ? options.dynamoGsi
      : new DynamoGsi(resource.dynamoTable);

    const defaultOptions: Required<AccessPatternOptions> = {
      dynamoGsi,
      pkAttributeOptions: {
        name: dynamoGsi.pkName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      skAttributeOptions: {
        name: dynamoGsi.skName,
        isPublic: false,
        generator: ResourceAttributeGenerator.COMPOSITION,
        generateOn: [
          OPERATION_SUB_TYPE.CREATE_ONE,
          OPERATION_SUB_TYPE.READ_ONE,
          OPERATION_SUB_TYPE.UPDATE_ONE,
          OPERATION_SUB_TYPE.DELETE_ONE,
        ],
      },
      type: ACCESS_PATTERN_TYPE.OTHER,
    };

    /***************************************************************************
     *
     * INIT ACCESS PATTERN
     *
     **************************************************************************/

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<AccessPatternOptions>;
    // the attribute might already be defined.
    const pkAttribute = resource.getAttributeByName(dynamoGsi.pkName);
    const skAttribute = resource.getAttributeByName(dynamoGsi.skName);

    // member components
    this.pkAttribute = pkAttribute
      ? pkAttribute
      : new ResourceAttribute(resource, this.options.pkAttributeOptions);
    this.skAttribute = skAttribute
      ? skAttribute
      : new ResourceAttribute(resource, this.options.skAttributeOptions);

    // parent + inverse
    this.resource = resource;
    this.resource.accessPatterns.push(this);

    this.project.logger.info(
      `Access Pattern for: "${this.pkAttribute.name}:${this.skAttribute.name}" initialized.`
    );

    return this;
  }

  public build() {
    this.project.logger.debug(
      `BUILD Access Pattern: "${this.pkAttribute.name}:${this.skAttribute.name}" called.`
    );
    /***************************************************************************
     *
     * CRUD OPERATIONS
     *
     * TODO
     *
     **************************************************************************/
    /*
    new Operation(this, {
      operationType: OPERATION_TYPE.MUTATION,
      operationSubType: OPERATION_SUB_TYPE.LIST,
    });
    */
  }

  addPkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.pkAttribute.compositionSources.push(sourceAttribute);
  }

  addSkAttributeSource(sourceAttribute: ResourceAttribute) {
    this.skAttribute.compositionSources.push(sourceAttribute);
  }

  /**
   * Returns operation for given sub-type
   */
  public getOperation(
    operationSubType: ValueOf<typeof OPERATION_SUB_TYPE>
  ): Operation {
    const returnOperation = this.operations.find(
      (o) => o.operationSubType === operationSubType
    );

    if (!returnOperation) {
      throw new Error(`Operation not found for sub-type: ${operationSubType}`);
    }

    return returnOperation;
  }

  public get createOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.CREATE_ONE);
  }

  public get readOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.READ_ONE);
  }

  public get updateOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.UPDATE_ONE);
  }

  public get deleteOperation(): Operation {
    return this.getOperation(OPERATION_SUB_TYPE.DELETE_ONE);
  }

  public get isKeyAccessPattern(): boolean {
    return this.dynamoTable.keyDynamoGsi === this.dynamoGsi;
  }

  public get isLookupAccessPattern(): boolean {
    return this.dynamoTable.lookupDynamoGsi === this.dynamoGsi;
  }

  public get dynamoTable(): DynamoTable {
    return this.dynamoGsi.dynamoTable;
  }

  public get dynamoGsi(): DynamoGsi {
    return this.options.dynamoGsi;
  }
}
