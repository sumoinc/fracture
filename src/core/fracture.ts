import { Component, LoggerOptions, Project, LogLevel } from "projen";
import { deepMerge } from "projen/lib/util";
import { AuditStrategy } from "./audit-strategy";
import { NamingStrategy, NAMING_STRATEGY_TYPE } from "./naming-strategy";
import { OPERATION_SUB_TYPE } from "./operation";
import { Organization, OrganizationOptions } from "./organization";
import {
  ResourceAttributeGenerator,
  ResourceAttributeType,
} from "./resource-attribute";
import { Service, ServiceOptions } from "./service";
import { STRUCTURE_TYPE } from "./structure";
import { TypescriptService } from "../generators/ts/typescript-service";

/**
 *
 * Top level options for Fracture. These options also apply to services and
 * other sub components as they are created, unless they are overridden.
 *
 */
export interface FractureOptions {
  /**
   * Directory where generated code will be placed.
   * @default project.outdir + "/" + namespace
   */
  outdir?: string;
  /**
   * Logging options
   * @default LogLevel.INFO
   */
  logging?: LoggerOptions;
  /**
   * Versioned.
   * @default true
   */
  isVersioned?: boolean;
  /**
   * The naming strategy to use for generated code.
   */
  namingStrategy?: NamingStrategy;
  /**
   * The audit strategy to use for generated code.
   */
  auditStrategy?: AuditStrategy;
}

/**
 * The root of the entire application.
 */
export class Fracture extends Component {
  // member components
  public readonly services: Service[];
  public readonly organizations: Organization[];
  // project and namespace
  public readonly project: Project;
  public readonly namespace: string;
  // all other options
  public readonly options: Required<FractureOptions>;

  constructor(
    project: Project,
    namespace: string = "fracture",
    options: FractureOptions = {}
  ) {
    super(project);

    /***************************************************************************
     *
     * DEFAULT OPTIONS
     *
     * These are the options that will be used through all code generation
     * unless explicitly overridden.
     *
     **************************************************************************/

    const defaultOptions: Required<FractureOptions> = {
      outdir: namespace,
      logging: {
        level: LogLevel.INFO,
      },
      isVersioned: true,
      namingStrategy: {
        ts: {
          attributeName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          className: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          enumName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          fileName: NAMING_STRATEGY_TYPE.PARAM_CASE,
          functionName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          functionParameterName: NAMING_STRATEGY_TYPE.CAMEL_CASE,
          interfaceName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
          typeName: NAMING_STRATEGY_TYPE.PASCAL_CASE,
        },
        appsync: {
          vtl: {
            file: NAMING_STRATEGY_TYPE.PARAM_CASE,
          },
        },
        attributes: {
          compositionSeperator: "#",
        },
        structures: {
          prefixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "",
            [STRUCTURE_TYPE.OUTPUT]: "",
            [STRUCTURE_TYPE.TRANSIENT]: "",
          },
          suffixes: {
            [STRUCTURE_TYPE.DATA]: "",
            [STRUCTURE_TYPE.INPUT]: "input",
            [STRUCTURE_TYPE.OUTPUT]: "output",
            [STRUCTURE_TYPE.TRANSIENT]: "message",
          },
        },
        operations: {
          prefixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "create",
            [OPERATION_SUB_TYPE.READ_ONE]: "get",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "update",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "delete",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "import",
            [OPERATION_SUB_TYPE.LIST]: "list",
          },
          suffixes: {
            [OPERATION_SUB_TYPE.CREATE_ONE]: "",
            [OPERATION_SUB_TYPE.READ_ONE]: "",
            [OPERATION_SUB_TYPE.UPDATE_ONE]: "",
            [OPERATION_SUB_TYPE.DELETE_ONE]: "",
            [OPERATION_SUB_TYPE.IMPORT_ONE]: "",
            [OPERATION_SUB_TYPE.LIST]: "",
          },
        },
      },
      auditStrategy: {
        create: {
          dateAttribute: {
            name: "created-at",
            shortName: "cd",
            comments: [`The date and time this record was created.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.CREATE_ONE],
          },
        },
        update: {
          dateAttribute: {
            name: "updated-at",
            shortName: "ud",
            comments: [`The date and time this record was last updated.`],
            type: ResourceAttributeType.DATE_TIME,
            isRequired: true,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [
              OPERATION_SUB_TYPE.CREATE_ONE,
              OPERATION_SUB_TYPE.UPDATE_ONE,
              OPERATION_SUB_TYPE.DELETE_ONE,
            ],
          },
        },
        delete: {
          dateAttribute: {
            name: "deleted-at",
            shortName: "dd",
            comments: [`The date and time this record was deleted.`],
            type: ResourceAttributeType.DATE_TIME,
            generator: ResourceAttributeGenerator.CURRENT_DATE_TIME_STAMP,
            generateOn: [OPERATION_SUB_TYPE.DELETE_ONE],
            outputOn: [OPERATION_SUB_TYPE.DELETE_ONE],
          },
        },
      },
    };

    /***************************************************************************
     *
     * INIT FRACTURE PROJECT
     *
     **************************************************************************/

    // member components
    this.services = [];
    this.organizations = [];

    // project and namespace
    this.project = project;
    this.namespace = namespace;

    // all other options
    this.options = deepMerge([
      defaultOptions,
      options,
    ]) as Required<FractureOptions>;

    return this;
  }

  /**
   * Build the project.
   *
   * Call this when you've configured everything, prior to preSynthesize.
   *
   * @returns void
   */
  public build() {
    this.services.forEach((service) => {
      service.build();
      new TypescriptService(service);
    });
  }

  /*****************************************************************************
   *
   *  Configuration Helpers
   *
   ****************************************************************************/

  /**
   * Add a service to the fracture project.
   * @param {ServiceOptions}
   * @returns {Service}
   */
  public addService(options: ServiceOptions) {
    return new Service(this, options);
  }

  /**
   * Add an organization to the fracture project.
   * @param {OrganizationOptions}
   * @returns {Organization}
   */
  public addOrganization(options: OrganizationOptions) {
    return new Organization(this, options);
  }
}
