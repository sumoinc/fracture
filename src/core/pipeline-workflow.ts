import { Component } from "projen";
import { GitHub, GithubWorkflow } from "projen/lib/github";
import { Fracture } from "./fracture";
import { Pipeline } from "./pipeline";

// const BUILD_JOBID = "build";
// const DEPLOY_JOBID_PREFIX = "deploy";
export const DEPLOY_TASK_PREFIX = "cdk:deploy";

export interface PipelineWorkflowOptions {
  pipeline: Pipeline;
}

export class PipelineWorkflow extends Component {
  // all other options
  public readonly options: PipelineWorkflowOptions;

  /**
   * Because the github workflow is build on the base project we pass in the app
   * but the component is actually based the root project
   */
  constructor(fracture: Fracture, options: PipelineWorkflowOptions) {
    super(fracture);

    this.options = options;

    // debugging output
    this.project.logger.info(
      `INIT Pipeline Workflow: "${this.pipeline.deployName}"`
    );
  }

  public get pipeline() {
    return this.options.pipeline;
  }

  preSynthesize(): void {
    const fracture = this.project as Fracture;
    const pipeline = this.pipeline;
    const app = pipeline.app;

    /*******************************************************************************
     *
     * TASK CONFIGURATION
     *
     * Adds a task to projen for each deployment stage.
     *
     ******************************************************************************/

    pipeline.stages.forEach((stage) => {
      fracture.addTask(`${DEPLOY_TASK_PREFIX}:${stage.id}`, {
        description: `CDK deploy of ${stage.name} stage`,
        exec: `cdk deploy *-${stage.id} --require-approval never`,
      });
    });

    /*******************************************************************************
     *
     * WORKFLOW CONFIGURATION
     *
     * Builds github workflow for stages created in tasks above.
     *
     ******************************************************************************/

    console.log(`DEBUG: "${pipeline.deployName}"`, fracture.name, app.name);

    const github = GitHub.of(fracture);
    if (!github) {
      throw new Error("Pipeline requires a github project");
    }

    new GithubWorkflow(github, pipeline.deployName);

    // build code

    // new BuildWorkflow(project, {
    //   name: "deploy",
    //   buildTask: project.buildTask,
    //   artifactsDirectory: project.artifactsDirectory,
    // containerImage: options.workflowContainerImage,
    // gitIdentity: this.workflowGitIdentity,
    // mutableBuild: options.mutableBuild,
    // preBuildSteps: this.renderWorkflowSetup({
    //   mutable: options.mutableBuild ?? true,
    // }),
    // postBuildSteps: options.postBuildSteps,
    // runsOn: options.workflowRunsOn,
    // workflowTriggers: options.buildWorkflowTriggers,
    // permissions: workflowPermissions,
    // });
  }
}
