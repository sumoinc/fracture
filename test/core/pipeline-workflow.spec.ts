import { synthSnapshot } from "projen/lib/util/synth";
import { Fracture } from "../../src/core/fracture";
import { DEPLOY_TASK_PREFIX } from "../../src/core/pipeline-workflow";
import { TestFracture } from "../util";

describe("Task tests", () => {
  test("Generates expected task", () => {
    const p = new TestFracture();
    const stage = p.testStage();

    const snapshot = synthSnapshot(p);

    console.log(snapshot[".projen/tasks.json"].tasks);

    expect(
      snapshot[".projen/tasks.json"].tasks[
        `${DEPLOY_TASK_PREFIX}:${stage.name}`
      ]
    ).toBeDefined();
  });
});

describe.only("workflow file tests", () => {
  test("Generates expected pipeline", () => {
    const p = new TestFracture();
    const pipeline = p.testPipeline();
    p.testStage();

    const workflows = synthWorkflows(p);

    const fileContent =
      workflows[`.github/workflows/${pipeline.deployName}.yml`];
    console.log(workflows[`.github/workflows/${pipeline.deployName}.yml`]);
    expect(fileContent).toBeDefined();
    expect(fileContent).toMatchSnapshot();
  });
});

function synthWorkflows(p: Fracture): any {
  const snapshot = synthSnapshot(p);
  const filtered = Object.keys(snapshot)
    .filter((path) => path.startsWith(".github/workflows/"))
    .reduce((obj, key) => {
      return {
        ...obj,
        [key]: snapshot[key],
      };
    }, {});
  return filtered;
}
