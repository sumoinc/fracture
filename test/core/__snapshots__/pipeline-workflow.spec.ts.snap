// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`workflow file tests Generates expected pipeline 1`] = `
"# ~~ Generated by projen. To modify, edit .projenrc.ts and run \\"npx projen\\".

name: deploy-test-pipeline
on: {}
jobs: {}
"
`;
